# Housing Rental Analysis for San Francisco

In this challenge, your job is to use your data visualization skills, including aggregation, interactive visualizations, and geospatial analysis, to find properties in the San Francisco market that are viable investment opportunities.

## Instructions

Use the `san_francisco_housing.ipynb` notebook to visualize and analyze the real-estate data.

Note that this assignment requires you to create a visualization by using hvPlot and GeoViews. Additionally, you need to read the `sfo_neighborhoods_census_data.csv` file from the `Resources` folder into the notebook and create the DataFrame that you’ll use in the analysis.

The main task in this Challenge is to visualize and analyze the real-estate data in your Jupyter notebook. Use the `san_francisco_housing.ipynb` notebook to complete the following tasks:

* Calculate and plot the housing units per year.

* Calculate and plot the average prices per square foot.

* Compare the average prices by neighborhood.

* Build an interactive neighborhood map.

* Compose your data story.

### Calculate and Plot the Housing Units per Year

For this part of the assignment, use numerical and visual aggregation to calculate the number of housing units per year, and then visualize the results as a bar chart. To do so, complete the following steps:

1. Use the `groupby` function to group the data by year. Aggregate the results by the `mean` of the groups.

2. Use the `hvplot` function to plot the `housing_units_by_year` DataFrame as a bar chart. Make the x-axis represent the `year` and the y-axis represent the `housing_units`.

3. Style and format the line plot to ensure a professionally styled visualization.

4. Note that your resulting plot should appear similar to the following image:

![A screenshot depicts an example of the resulting bar chart.](Images/zoomed-housing-units-by-year.png)

5. Answer the following question:

    * What’s the overall trend in housing units over the period that you’re analyzing?

### Calculate and Plot the Average Sale Prices per Square Foot

For this part of the assignment, use numerical and visual aggregation to calculate the average prices per square foot, and then visualize the results as a bar chart. To do so, complete the following steps:

1. Group the data by year, and then average the results. What’s the lowest gross rent that’s reported for the years that the DataFrame includes?

2. Create a new DataFrame named `prices_square_foot_by_year` by filtering out the “housing_units” column. The new DataFrame should include the averages per year for only the sale price per square foot and the gross rent.

3. Use hvPlot to plot the `prices_square_foot_by_year` DataFrame as a line plot.

    > **Hint** This single plot will include lines for both `sale_price_sqr_foot` and `gross_rent`.

4. Style and format the line plot to ensure a professionally styled visualization.

5. Note that your resulting plot should appear similar to the following image:

![A screenshot depicts an example of the resulting plot.](Images/avg-sale-px-sq-foot-gross-rent.png)

6. Use both the `prices_square_foot_by_year` DataFrame and interactive plots to answer the following questions:

    * Did any year experience a drop in the average sale price per square foot compared to the previous year?

    * If so, did the gross rent increase or decrease during that year?

### Compare the Average Sale Prices by Neighborhood

For this part of the assignment, use interactive visualizations and widgets to explore the average sale price per square foot by neighborhood. To do so, complete the following steps:

1. Create a new DataFrame that groups the original DataFrame by year and neighborhood. Aggregate the results by the `mean` of the groups.

2. Filter out the “housing_units” column to create a DataFrame that includes only the `sale_price_sqr_foot` and `gross_rent` averages per year.

3. Create an interactive line plot with hvPlot that visualizes both `sale_price_sqr_foot` and `gross_rent`. Set the x-axis parameter to the year (`x="year"`). Use the `groupby` parameter to create an interactive widget for `neighborhood`.

4. Style and format the line plot to ensure a professionally styled visualization.

5. Note that your resulting plot should appear similar to the following image:

![A screenshot depicts an example of the resulting plot.](Images/pricing-info-by-neighborhood.png)

6. Use the interactive visualization to answer the following question:

    * For the Anza Vista neighborhood, is the average sale price per square foot for 2016 more or less than the price that’s listed for 2012? 

### Build an Interactive Neighborhood Map

For this part of the assignment, explore the geospatial relationships in the data by using interactive visualizations with hvPlot and GeoViews. To build your map, use the `sfo_data_df` DataFrame (created during the initial import), which includes the neighborhood location data with the average prices. To do all this, complete the following steps:

1. Read the `neighborhood_coordinates.csv` file from the `Resources` folder into the notebook, and create a DataFrame named `neighborhood_locations_df`. Be sure to set the `index_col` of the DataFrame as “Neighborhood”.

2. Using the original `sfo_data_df` Dataframe, create a DataFrame named `all_neighborhood_info_df` that groups the data by neighborhood. Aggregate the results by the `mean` of the group.

3. Review the two code cells that concatenate the `neighborhood_locations_df` DataFrame with the `all_neighborhood_info_df` DataFrame. Note that the first cell uses the [Pandas concat function](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.concat.html) to create a DataFrame named `all_neighborhoods_df`. The second cell cleans the data and sets the “Neighborhood” column. Be sure to run these cells to create the `all_neighborhoods_df` DataFrame, which you’ll need to create the geospatial visualization.

4. Using hvPlot with GeoViews enabled, create a `points` plot for the `all_neighborhoods_df` DataFrame. Be sure to do the following:

    * Set the `geo` parameter to True.
    * Set the `size` parameter to “sale_price_sqr_foot”.
    * Set the `color` parameter to “gross_rent”.
    * Set the `frame_width` parameter to 700.
    * Set the `frame_height` parameter to 500.
    * Include a descriptive title.

Note that your resulting plot should appear similar to the following image:

![A screenshot depicts an example of a scatter plot created with hvPlot and GeoViews.](Images/6-4-geoviews-plot.png)

5. Use the interactive map to answer the following question:

    * Which neighborhood has the highest gross rent, and which has the highest sale price per square foot?

### Compose Your Data Story

Based on the visualizations that you created, answer the following questions:

* How does the trend in rental income growth compare to the trend in sales prices? Does this same trend hold true for all the neighborhoods across San Francisco?

* What insights can you share with your company about the potential one-click, buy-and-rent strategy that they're pursuing? Do neighborhoods exist that you would suggest for investment, and why?

In [229]:
# Import the required libraries and dependencies
import pandas as pd
import hvplot.pandas
from pathlib import Path
import panel as pn

## Import the data 

In [230]:
# Using the read_csv function and Path module, create a DataFrame 
# by importing the sfo_neighborhoods_census_data.csv file from the Resources folder
sfo_data_df = pd.read_csv(Path("Resources/sfo_neighborhoods_census_data.csv"), index_col= "year")
sfo_data_df.dropna()

# Review the first five rows of the DataFrame
display(sfo_data_df.head())
display(sfo_data_df.tail())

neighborhood  sale_price_sqr_foot  housing_units  gross_rent
year                                                                  
2010      Alamo Square           291.182945         372560        1239
2010        Anza Vista           267.932583         372560        1239
2010           Bayview           170.098665         372560        1239
2010  Buena Vista Park           347.394919         372560        1239
2010  Central Richmond           319.027623         372560        1239

neighborhood  sale_price_sqr_foot  housing_units  gross_rent
year                                                                        
2016          Telegraph Hill           903.049771         384242        4390
2016              Twin Peaks           970.085470         384242        4390
2016  Van Ness/ Civic Center           552.602567         384242        4390
2016       Visitacion Valley           328.319007         384242        4390
2016           Westwood Park           631.195426         384242        4390

---

## Calculate and Plot the Housing Units per Year

For this part of the assignment, use numerical and visual aggregation to calculate the number of housing units per year, and then visualize the results as a bar chart. To do so, complete the following steps:

1. Use the `groupby` function to group the data by year. Aggregate the results by the `mean` of the groups.

2. Use the `hvplot` function to plot the `housing_units_by_year` DataFrame as a bar chart. Make the x-axis represent the `year` and the y-axis represent the `housing_units`.

3. Style and format the line plot to ensure a professionally styled visualization.

4. Note that your resulting plot should appear similar to the following image:

![A screenshot depicts an example of the resulting bar chart.](Images/zoomed-housing-units-by-year.png)

5. Answer the following question:

    * What’s the overall trend in housing units over the period that you’re analyzing?



### Step 1: Use the `groupby` function to group the data by year. Aggregate the results by the `mean` of the groups.

In [231]:
# Create a numerical aggregation that groups the data by the year and then averages the results.
housing_units_by_year = (sfo_data_df[["housing_units"]].groupby("year").mean().sort_values("year"))        # mean number of units       
housing_units_by_year_error = (sfo_data_df[["housing_units"]].groupby("year").sem().sort_values("year"))    # Just curious about error     

# Review the DataFrame
display(housing_units_by_year)
display(housing_units_by_year_error)                                      # Hmmm....

housing_units
year               
2010       372560.0
2011       374507.0
2012       376454.0
2013       378401.0
2014       380348.0
2015       382295.0
2016       384242.0

housing_units
year               
2010            0.0
2011            0.0
2012            0.0
2013            0.0
2014            0.0
2015            0.0
2016            0.0

### Step 2: Use the `hvplot` function to plot the `housing_units_by_year` DataFrame as a bar chart. Make the x-axis represent the `year` and the y-axis represent the `housing_units`.

### Step 3: Style and format the line plot to ensure a professionally styled visualization.

In [232]:
# Create a visual aggregation explore the housing units by year
mean_hu_x_yr = housing_units_by_year.hvplot.bar(y = "housing_units",                                      # y-axis data
                                 x = "year",                                               # x-axis data
                                 xlabel = "Year",                                          # x-axis label
                                 ylim = (365000, 386000),                                  # y-axis scaling for clarity of data
                                ylabel = "Housing units",                                  # y-axis label
                                 height = 600,                                             # set vertical dimension of plot
                                 width = 900,                                              # set horizontal dimension of plot
                                 title = "Housing units in San Francisco 2010 to 2016",    # giving the plot a descriptive title
                                ).opts(yformatter="%.0f",                                  # un-did scientific notation, no decimal
                                       bgcolor= "lightgrey",                               # I thought it would look good
                                      hover_color= "orange",                               # for presentation purposes
                                      fontsize= {"title": 24,                              # Getting the font sizes a little more
                                                 "labels": 36,                             # more useful for reading in print
                                                 "xticks": 18,
                                                 "yticks": 18})

mean_hu_x_yr                                                                               # Assigned to a variable for potential
                                                                                           # later use. Called here.

:Bars   [year]   (housing_units)

### Step 5: Answer the following question:

**Question:** What is the overall trend in housing_units over the period being analyzed?

# **Answer:** The number of housing units has increased by approximately 40% over the period from 2010 to 2016. The increase appears to have been a steady linear function. Bars indicate the mean number of housing units. There is curiously no error.

---

## Calculate and Plot the Average Sale Prices per Square Foot

For this part of the assignment, use numerical and visual aggregation to calculate the average prices per square foot, and then visualize the results as a bar chart. To do so, complete the following steps:

1. Group the data by year, and then average the results. What’s the lowest gross rent that’s reported for the years that the DataFrame includes?

2. Create a new DataFrame named `prices_square_foot_by_year` by filtering out the “housing_units” column. The new DataFrame should include the averages per year for only the sale price per square foot and the gross rent.

3. Use hvPlot to plot the `prices_square_foot_by_year` DataFrame as a line plot.

    > **Hint** This single plot will include lines for both `sale_price_sqr_foot` and `gross_rent`.

4. Style and format the line plot to ensure a professionally styled visualization.

5. Note that your resulting plot should appear similar to the following image:

![A screenshot depicts an example of the resulting plot.](Images/avg-sale-px-sq-foot-gross-rent.png)

6. Use both the `prices_square_foot_by_year` DataFrame and interactive plots to answer the following questions:

    * Did any year experience a drop in the average sale price per square foot compared to the previous year?

    * If so, did the gross rent increase or decrease during that year?



### Step 1: Group the data by year, and then average the results.

In [233]:
# Create a numerical aggregation by grouping the data by year and averaging the results
prices_square_foot_by_year = (sfo_data_df[["sale_price_sqr_foot"]].groupby("year").mean())       # mean price/ sq ft
prices_square_foot_by_year_error = (sfo_data_df[["sale_price_sqr_foot"]].groupby("year").sem())    # sem price/ sq ft

# Review the resulting DataFrame
display(prices_square_foot_by_year)                                                            # displays the means
display(prices_square_foot_by_year_error)                                                      # displays the sems

sale_price_sqr_foot
year                     
2010           369.344353
2011           341.903429
2012           399.389968
2013           483.600304
2014           556.277273
2015           632.540352
2016           697.643709

sale_price_sqr_foot
year                     
2010            20.230206
2011            13.448431
2012            14.939303
2013            25.002277
2014            26.202415
2015            40.701298
2016            47.947735

**Question:** What is the lowest gross rent reported for the years included in the DataFrame?

# **Answer:** The lowest gross rent reported for the period of 2010 to 2016 was 341.90 +/- 13.45 (mean +/- SEM, in USD).

### Step 2: Create a new DataFrame named `prices_square_foot_by_year` by filtering out the “housing_units” column. The new DataFrame should include the averages per year for only the sale price per square foot and the gross rent.

In [234]:
# Filter out the housing_units column, creating a new DataFrame 
# Keep only sale_price_sqr_foot and gross_rent averages per year
prices_square_foot_by_year = (sfo_data_df[["sale_price_sqr_foot", "gross_rent"]].groupby("year").mean())
prices_square_foot_by_year_error = (sfo_data_df[["sale_price_sqr_foot", "gross_rent"]].groupby("year").sem())

# Review the DataFrame
display(prices_square_foot_by_year)
display(prices_square_foot_by_year_error)

sale_price_sqr_foot  gross_rent
year                                 
2010           369.344353      1239.0
2011           341.903429      1530.0
2012           399.389968      2324.0
2013           483.600304      2971.0
2014           556.277273      3528.0
2015           632.540352      3739.0
2016           697.643709      4390.0

sale_price_sqr_foot  gross_rent
year                                 
2010            20.230206         0.0
2011            13.448431         0.0
2012            14.939303         0.0
2013            25.002277         0.0
2014            26.202415         0.0
2015            40.701298         0.0
2016            47.947735         0.0

### Step 3: Use hvPlot to plot the `prices_square_foot_by_year` DataFrame as a line plot.

> **Hint** This single plot will include lines for both `sale_price_sqr_foot` and `gross_rent`

### Step 4: Style and format the line plot to ensure a professionally styled visualization.


In [237]:
# Plot prices_square_foot_by_year. 
# Inclued labels for the x- and y-axes, and a title.
plot_a = prices_square_foot_by_year.hvplot.line(
    xlabel= "Year",
    ylabel= "Gross rent/ Sale price per square foot",
    height= 700,
    width= 1000,
    title= "Mean sale price per square foot & Mean gross rent for San Fransisco, 2010 to 2016"
    ).opts(bgcolor= "lightgrey",
           fontsize= {"title": 16, "xlabel":24, "ylabel": 24, "xticks": 14, "yticks": 14, "legend":14, "legend_title":16},
          legend_offset= (0, 500),
          )

plot_a

:NdOverlay   [Variable]
   :Curve   [year]   (value)

### Step 6: Use both the `prices_square_foot_by_year` DataFrame and interactive plots to answer the following questions:

**Question:** Did any year experience a drop in the average sale price per square foot compared to the previous year?

# **Answer:** Yes, the average sale price per square foot was lower in 2011 than it was in 2010 (blue line in figure).

**Question:** If so, did the gross rent increase or decrease during that year?

# **Answer:** Although the sale price per square foot decreased in 2011, the gross rent increased during that year.

---

## Compare the Average Sale Prices by Neighborhood

For this part of the assignment, use interactive visualizations and widgets to explore the average sale price per square foot by neighborhood. To do so, complete the following steps:

1. Create a new DataFrame that groups the original DataFrame by year and neighborhood. Aggregate the results by the `mean` of the groups.

2. Filter out the “housing_units” column to create a DataFrame that includes only the `sale_price_sqr_foot` and `gross_rent` averages per year.

3. Create an interactive line plot with hvPlot that visualizes both `sale_price_sqr_foot` and `gross_rent`. Set the x-axis parameter to the year (`x="year"`). Use the `groupby` parameter to create an interactive widget for `neighborhood`.

4. Style and format the line plot to ensure a professionally styled visualization.

5. Note that your resulting plot should appear similar to the following image:

![A screenshot depicts an example of the resulting plot.](Images/pricing-info-by-neighborhood.png)

6. Use the interactive visualization to answer the following question:

    * For the Anza Vista neighborhood, is the average sale price per square foot for 2016 more or less than the price that’s listed for 2012? 


### Step 1: Create a new DataFrame that groups the original DataFrame by year and neighborhood. Aggregate the results by the `mean` of the groups.

In [238]:
# Group by year and neighborhood and then create a new dataframe of the mean values
prices_by_year_by_neighborhood = sfo_data_df.groupby(["year", "neighborhood"]).mean()

# Review the DataFrame
display(prices_by_year_by_neighborhood.head())
display(prices_by_year_by_neighborhood.tail())

sale_price_sqr_foot  housing_units  gross_rent
year neighborhood                                                    
2010 Alamo Square               291.182945       372560.0      1239.0
     Anza Vista                 267.932583       372560.0      1239.0
     Bayview                    170.098665       372560.0      1239.0
     Buena Vista Park           347.394919       372560.0      1239.0
     Central Richmond           319.027623       372560.0      1239.0

sale_price_sqr_foot  housing_units  gross_rent
year neighborhood                                                          
2016 Telegraph Hill                   903.049771       384242.0      4390.0
     Twin Peaks                       970.085470       384242.0      4390.0
     Van Ness/ Civic Center           552.602567       384242.0      4390.0
     Visitacion Valley                328.319007       384242.0      4390.0
     Westwood Park                    631.195426       384242.0      4390.0

### Step 2: Filter out the “housing_units” column to create a DataFrame that includes only the `sale_price_sqr_foot` and `gross_rent` averages per year.

In [239]:
# Filter out the housing_units
prices_by_year_by_neighborhood = prices_by_year_by_neighborhood[["sale_price_sqr_foot", "gross_rent"]]

# Review the first and last five rows of the DataFrame
display(prices_by_year_by_neighborhood.head())
display(prices_by_year_by_neighborhood.tail())

sale_price_sqr_foot  gross_rent
year neighborhood                                     
2010 Alamo Square               291.182945      1239.0
     Anza Vista                 267.932583      1239.0
     Bayview                    170.098665      1239.0
     Buena Vista Park           347.394919      1239.0
     Central Richmond           319.027623      1239.0

sale_price_sqr_foot  gross_rent
year neighborhood                                           
2016 Telegraph Hill                   903.049771      4390.0
     Twin Peaks                       970.085470      4390.0
     Van Ness/ Civic Center           552.602567      4390.0
     Visitacion Valley                328.319007      4390.0
     Westwood Park                    631.195426      4390.0

### Step 3: Create an interactive line plot with hvPlot that visualizes both `sale_price_sqr_foot` and `gross_rent`. Set the x-axis parameter to the year (`x="year"`). Use the `groupby` parameter to create an interactive widget for `neighborhood`.

### Step 4: Style and format the line plot to ensure a professionally styled visualization.

In [241]:
# Use hvplot to create an interactive line plot of the average price per square foot
# The plot should have a dropdown selector for the neighborhood
plot_b = prices_by_year_by_neighborhood.hvplot(groupby="neighborhood",
    x="year",
    xlabel= "Year",
    ylabel= "Gross rent/ Sale price per square foot",
    height= 700,
    width= 1000,
    title= "Mean sale price per square foot & Mean gross rent for San Fransisco, 2010 to 2016"
    ).opts(bgcolor= "lightgrey",
           fontsize= {"title": 16, "xlabel":24, "ylabel": 24, "xticks": 14, "yticks": 14, "legend":14, "legend_title":16},
          legend_offset= (0, 500),
          )#.layout().cols(1)


plot_b

:DynamicMap   [neighborhood]
   :NdOverlay   [Variable]
      :Curve   [year]   (value)

### Step 6: Use the interactive visualization to answer the following question:

**Question:** For the Anza Vista neighborhood, is the average sale price per square foot for 2016 more or less than the price that’s listed for 2012? 

**Answer:** I would love to be able to answer this question but the graph does not change when I select Anza Vista from the 
dropdown. I downloaded examples from the hvplot user guide and they would not work properly either. So I just dropped the groupby and added .layout().cols(1) at the end and printed out all of the plots

Answer to the question:  For the Anza Vista neighborhood, the average sale price per square foot for 2016 was less than the price that is listed for 2012 (344.49 and 88.40 for 2016 and 2012, respectively).

I commented out the layout operation so that you can see what I did to get around the non-functional groupby dropdown. If you uncomment and run with the layout operation then the second figure produced is for Anza Vista.

---

## Build an Interactive Neighborhood Map

For this part of the assignment, explore the geospatial relationships in the data by using interactive visualizations with hvPlot and GeoViews. To build your map, use the `sfo_data_df` DataFrame (created during the initial import), which includes the neighborhood location data with the average prices. To do all this, complete the following steps:

1. Read the `neighborhood_coordinates.csv` file from the `Resources` folder into the notebook, and create a DataFrame named `neighborhood_locations_df`. Be sure to set the `index_col` of the DataFrame as “Neighborhood”.

2. Using the original `sfo_data_df` Dataframe, create a DataFrame named `all_neighborhood_info_df` that groups the data by neighborhood. Aggregate the results by the `mean` of the group.

3. Review the two code cells that concatenate the `neighborhood_locations_df` DataFrame with the `all_neighborhood_info_df` DataFrame. Note that the first cell uses the [Pandas concat function](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.concat.html) to create a DataFrame named `all_neighborhoods_df`. The second cell cleans the data and sets the “Neighborhood” column. Be sure to run these cells to create the `all_neighborhoods_df` DataFrame, which you’ll need to create the geospatial visualization.

4. Using hvPlot with GeoViews enabled, create a `points` plot for the `all_neighborhoods_df` DataFrame. Be sure to do the following:

    * Set the `size` parameter to “sale_price_sqr_foot”.

    * Set the `color` parameter to “gross_rent”.

    * Set the `size_max` parameter to “25”.

    * Set the `zoom` parameter to “11”.

Note that your resulting plot should appear similar to the following image:

![A screenshot depicts an example of a scatter plot created with hvPlot and GeoViews.](Images/6-4-geoviews-plot.png)

5. Use the interactive map to answer the following question:

    * Which neighborhood has the highest gross rent, and which has the highest sale price per square foot?


### Step 1: Read the `neighborhood_coordinates.csv` file from the `Resources` folder into the notebook, and create a DataFrame named `neighborhood_locations_df`. Be sure to set the `index_col` of the DataFrame as “Neighborhood”.

In [242]:
# Load neighborhoods coordinates data
neighborhood_locations_df = pd.read_csv(Path("Resources/neighborhoods_coordinates.csv"), index_col= "Neighborhood")

# Review the DataFrame
display(neighborhood_locations_df.head())
display(neighborhood_locations_df.tail())

Lat         Lon
Neighborhood                          
Alamo Square     37.791012 -122.402100
Anza Vista       37.779598 -122.443451
Bayview          37.734670 -122.401060
Bayview Heights  37.728740 -122.410980
Bernal Heights   37.728630 -122.443050

Lat         Lon
Neighborhood                            
West Portal         37.74026 -122.463880
Western Addition    37.79298 -122.435790
Westwood Highlands  37.73470 -122.456854
Westwood Park       37.73415 -122.457000
Yerba Buena         37.79298 -122.396360

### Step 2: Using the original `sfo_data_df` Dataframe, create a DataFrame named `all_neighborhood_info_df` that groups the data by neighborhood. Aggregate the results by the `mean` of the group.

In [243]:
# Calculate the mean values for each neighborhood
all_neighborhood_info_df = sfo_data_df.groupby("neighborhood").mean()

# Review the resulting DataFrame
display(all_neighborhood_info_df.head())
display(all_neighborhood_info_df.tail())

sale_price_sqr_foot  housing_units   gross_rent
neighborhood                                                    
Alamo Square              366.020712       378401.0  2817.285714
Anza Vista                373.382198       379050.0  3031.833333
Bayview                   204.588623       376454.0  2318.400000
Bayview Heights           590.792839       382295.0  3739.000000
Bernal Heights            576.746488       379374.5  3080.333333

sale_price_sqr_foot  housing_units   gross_rent
neighborhood                                                       
West Portal                  498.488485      376940.75  2515.500000
Western Addition             307.562201      377427.50  2555.166667
Westwood Highlands           533.703935      376454.00  2250.500000
Westwood Park                687.087575      382295.00  3959.000000
Yerba Buena                  576.709848      377427.50  2555.166667

### Step 3: Review the two code cells that concatenate the `neighborhood_locations_df` DataFrame with the `all_neighborhood_info_df` DataFrame. 

Note that the first cell uses the [Pandas concat function](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.concat.html) to create a DataFrame named `all_neighborhoods_df`. 

The second cell cleans the data and sets the “Neighborhood” column. 

Be sure to run these cells to create the `all_neighborhoods_df` DataFrame, which you’ll need to create the geospatial visualization.

In [244]:
# Using the Pandas `concat` function, join the 
# neighborhood_locations_df and the all_neighborhood_info_df DataFrame
# The axis of the concatenation is "columns".
# The concat function will automatially combine columns with
# identical information, while keeping the additional columns.
all_neighborhoods_df = pd.concat(
    [neighborhood_locations_df, all_neighborhood_info_df], 
    axis="columns",
    sort=False
)

# Review the resulting DataFrame
display(all_neighborhoods_df.head())
display(all_neighborhoods_df.tail())


Lat         Lon  sale_price_sqr_foot  housing_units  \
Alamo Square     37.791012 -122.402100           366.020712       378401.0   
Anza Vista       37.779598 -122.443451           373.382198       379050.0   
Bayview          37.734670 -122.401060           204.588623       376454.0   
Bayview Heights  37.728740 -122.410980           590.792839       382295.0   
Bernal Heights   37.728630 -122.443050                  NaN            NaN   

                  gross_rent  
Alamo Square     2817.285714  
Anza Vista       3031.833333  
Bayview          2318.400000  
Bayview Heights  3739.000000  
Bernal Heights           NaN

Lat        Lon  sale_price_sqr_foot  housing_units  \
Yerba Buena      37.79298 -122.39636           576.709848       377427.5   
Bernal Heights        NaN        NaN           576.746488       379374.5   
Downtown              NaN        NaN           391.434378       378401.0   
Ingleside             NaN        NaN           367.895144       377427.5   
Outer Richmond        NaN        NaN           473.900773       378401.0   

                  gross_rent  
Yerba Buena      2555.166667  
Bernal Heights   3080.333333  
Downtown         2817.285714  
Ingleside        2509.000000  
Outer Richmond   2817.285714

In [245]:
# Call the dropna function to remove any neighborhoods that do not have data
all_neighborhoods_df = all_neighborhoods_df.reset_index().dropna()

# Rename the "index" column as "Neighborhood" for use in the Visualization
all_neighborhoods_df = all_neighborhoods_df.rename(columns={"index": "Neighborhood"})

# Review the resulting DataFrame
display(all_neighborhoods_df.head())
display(all_neighborhoods_df.tail())

Neighborhood        Lat         Lon  sale_price_sqr_foot  \
0      Alamo Square  37.791012 -122.402100           366.020712   
1        Anza Vista  37.779598 -122.443451           373.382198   
2           Bayview  37.734670 -122.401060           204.588623   
3   Bayview Heights  37.728740 -122.410980           590.792839   
5  Buena Vista Park  37.768160 -122.439330           452.680591   

   housing_units   gross_rent  
0       378401.0  2817.285714  
1       379050.0  3031.833333  
2       376454.0  2318.400000  
3       382295.0  3739.000000  
5       378076.5  2698.833333

Neighborhood       Lat         Lon  sale_price_sqr_foot  \
68         West Portal  37.74026 -122.463880           498.488485   
69    Western Addition  37.79298 -122.435790           307.562201   
70  Westwood Highlands  37.73470 -122.456854           533.703935   
71       Westwood Park  37.73415 -122.457000           687.087575   
72         Yerba Buena  37.79298 -122.396360           576.709848   

    housing_units   gross_rent  
68      376940.75  2515.500000  
69      377427.50  2555.166667  
70      376454.00  2250.500000  
71      382295.00  3959.000000  
72      377427.50  2555.166667

### Step 4: Using hvPlot with GeoViews enabled, create a `points` plot for the `all_neighborhoods_df` DataFrame. Be sure to do the following:

* Set the `geo` parameter to True.
* Set the `size` parameter to “sale_price_sqr_foot”.
* Set the `color` parameter to “gross_rent”.
* Set the `frame_width` parameter to 700.
* Set the `frame_height` parameter to 500.
* Include a descriptive title.

In [246]:
# Create a plot to analyze neighborhood info
plot_c = all_neighborhoods_df.hvplot.points(
    'Lon',
    'Lat',
    geo=True,
    tiles='OSM',
    size= "sale_price_sqr_foot",
    color= "gross_rent",
    frame_width= 700,
    frame_height= 500,
    title="Mean sale price per square foot (point size) and Mean gross rent (color) for San Fransisco 2010 to 2016",
    hover_cols=["Neighborhood", "gross_rent", "sale_price_sqr_foot"]
)

plot_c

:Overlay
   .Tiles.I  :Tiles   [x,y]
   .Points.I :Points   [Lon,Lat]   (gross_rent,sale_price_sqr_foot,Neighborhood)

# In the figure above, mean sale price per square foot is represented by the size of the point with larger points represent higher sale price. Mean gross rent is represented by color of the point with darker colors representing greater mean gross rents.

### Step 5: Use the interactive map to answer the following question:

**Question:** Which neighborhood has the highest gross rent, and which has the highest sale price per square foot?

## **Answer:** Westwood park has the highest gross rent (3,959 USD). Union Square Park has the highest sale price per square foot (903.99 USD).

## Compose Your Data Story

Based on the visualizations that you have created, compose a data story that synthesizes your analysis by answering the following questions:

**Question:**  How does the trend in rental income growth compare to the trend in sales prices? Does this same trend hold true for all the neighborhoods across San Francisco?

## **Answer:** Analysis showed that rental income has grown considerably faster than has sales price. While rental income has increased steadily, sales prices in most neighborhoods have remained fairly flat over time. The trend of increased rent was true for all neighborhoods. It is not clear from our analysis exactly how much error is associated with the lines on the graph. Not having the error makes it impossible to guess whether the fluctuations in sales price are significant or not. I would like to see error bars or 95% confidence limits around the line in our graphs. The hypothesis that sale price has not changed significantly could be tested using a generalized linear model that tests for the main effects of year, neighborhood, and their interaction on sales price.

**Question:** What insights can you share with your company about the potential one-click, buy-and-rent strategy that they're pursuing? Do neighborhoods exist that you would suggest for investment, and why?

## **Answer:** The one-click, buy-and-rent strategy is ok for ballparking it however the company needs to be cautioned that mean values over 7 years are being used. Not knowing the error around those means makes it risky to just point and click select property. However, since sale price appeared to have remained relatively flat over the time period analyzed while rent increased dramatically, a strategy of selecting property based on higher rent could be successful.

## Goal higher sales price: From pointing and clicking it looks like neighborhoods with high rent and high sales price lie within the Miraloma Park, Sunnyside, Westwood Park cluster of neighborhoods and Bayview Heights, Excelsior neighborhoods.

## Goal higher rent: Union Square to Yurba Buena neighborhood region looked to have higher rents.